Engine/DataFeeds/Enumerators/QuoteBarBuilderEnumerator.cs

```diff
@@ -84,6 +84,13 @@ public void ProcessData(BaseData data)
                 var utcNow = _timeProvider.GetUtcNow();
                 var currentLocalTime = utcNow.ConvertFromUtc(_timeZone);
                 var barStartTime = currentLocalTime.RoundDown(_barSize);
+
+                if (barStartTime == currentLocalTime)
+                {
+                    // adjust start time if tick aligned to bar period
+                    barStartTime = currentLocalTime.Subtract(_barSize);
```

Comment:
Won't `barStartTime = currentLocalTime.Subtract(_barSize);` cause us to emit _right away_ (when queried), since `EndTime = barStartTime + _barSize = currentLocalTime - _barSize + _barSize = currentLocalTime`, doesn't seem right from the perspective of the `BuilderEnumerator`, since it won't repect bar size?

1 Hour bar size example of the issue:
- Initial tick (1) at 10:00 AM. Quote bar starts at 10:00 AM and ends at 11:00 AM
- Second tick (2) at 11:00 AM
- Race condition between removing the `working` quote bar (1) L124 and adding a new one or not L81, this is the issue I understand we are trying to fix, right @StefanoRaggi ?

Reply:
The issue found when writing the new tests in #3384 was that bars were emitted once every two bars (i.e. bar ending at 11:00 emitted, bar ending at 12:00 not emitted, bar ending at 13:00 emitted, etc.) and each bar emitted was including two ticks instead of one. For this reason, I tried to repeat the issue with a new unit test for this enumerator -- `AggregatesHourlyTicksIntoHourlyBars`, included in this PR.

Not sure if it's actually a race condition since `liveMode` is set to `false` in this test so the timer in the `RealTimeScheduleEventService` is disabled and `MoveNext` is only called by the test.

Reply:
Modified the `AggregatesHourlyTicksIntoHourlyBars` it passes in `master`: `.MoveNext()` before adding the new data point so we get the finished working bar and the new data point generates a new bar

```
       [Test]
        public void AggregatesHourlyTicksIntoHourlyBars()
        {
            var timeProvider = new ManualTimeProvider(TimeZones.NewYork);
            var enumerator = new TradeBarBuilderEnumerator(Time.OneHour, TimeZones.NewYork, timeProvider, false);

            // noon new york time
            var currentTime = new DateTime(2015, 10, 08, 12, 0, 0);
            timeProvider.SetCurrentTime(currentTime);

            var price = 0m;

            for (var i = 0; i < 5; i++)
            {
                // advance one hour
                currentTime = currentTime.AddHours(1);
                timeProvider.SetCurrentTime(currentTime);

                // we advanced time, a new bar should be generated
                Assert.IsTrue(enumerator.MoveNext());

                // the first loop no trade bar was generated yet, end time not reached
                if (i > 0)
                {
                    Assert.IsNotNull(enumerator.Current);

                    var bar = (TradeBar)enumerator.Current;
                    Assert.AreEqual(currentTime.AddHours(-1), bar.Time);
                    Assert.AreEqual(currentTime, bar.EndTime);
                    Assert.AreEqual(price, bar.Open);
                    Assert.AreEqual(price, bar.High);
                    Assert.AreEqual(price, bar.Low);
                    Assert.AreEqual(price, bar.Close);
                }

                // add a tick, will generate a new bar
                price++;
                enumerator.ProcessData(new Tick(currentTime, Symbols.SPY, price, price, price));
            }
```